src/coreclr/jit/flowgraph.cpp

```diff
@@ -52,6 +52,239 @@ static bool blockNeedsGCPoll(BasicBlock* block)
+//
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+
+PhaseStatus Compiler::fgInsertClsInitChecks()
+{
+    if (!opts.OptimizationEnabled())
+    {
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    if (!fgHasOptStaticInit)
+    {
+        JITDUMP("fgInsertClsInitChecks: compiler claims there are no cctors we can optimize.");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    bool        modified = false;
+    BasicBlock* block;
+    BasicBlock* prevBb = nullptr;
+    for (block = fgFirstBB; block; block = block->bbNext)
+    {
+        // Skip cold blocks. Also, we might want to split basic-blocks in the middle
+        // so let's skip those with BBF_GC_SAFE_POINT for simplicity.
+        if (!block->isRunRarely() && !(block->bbFlags & BBF_GC_SAFE_POINT))
+        {
+            Statement* prevStmt = nullptr;
+            for (Statement* stmt : block->Statements())
+            {
+                for (GenTree* tree = stmt->GetTreeList(); tree != nullptr; tree = tree->gtNext)
+                {
+                    // We only need GT_CALL nodes with helper funcs
+                    // Looks like BBF_HAS_CALL/GTF_CALL aren't reliable as a fast check.
+                    if (!tree->IsCall())
+                    {
+                        continue;
+                    }
+
+                    GenTreeCall* call = tree->AsCall();
+                    if ((call->gtCallType != CT_HELPER) || (call->gtRetClsHnd == nullptr))
+                    {
+                        continue;
+                    }
+
+                    CorInfoHelpFunc helpFunc = eeGetHelperNum(call->gtCallMethHnd);
+                    if ((helpFunc != CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE) &&
+                        (helpFunc != CORINFO_HELP_CLASSINIT_SHARED_DYNAMICCLASS))
+                    {
+                        continue;
+                    }
+                    assert(call->fgArgInfo->ArgCount() == 2);
+
+                    GenTree* moduleIdArg = call->fgArgInfo->GetArgNode(0);
+                    GenTree* clsIdArg    = call->fgArgInfo->GetArgNode(1);
+
+                    if (!moduleIdArg->IsCnsIntOrI() || !clsIdArg->IsCnsIntOrI())
+                    {
+                        // ModuleId or/and clsId were passed as indirect loads
+                        // We can consider optimizing this case too (for R2R)
+                        // but it most likely will come with a noticeable size regression.
+                        continue;
+                    }
+
+                    int    isInitMask = 0;
+                    size_t isInitAddr = info.compCompHnd->getIsClassInitedFieldAddress(call->gtRetClsHnd, &isInitMask);
+                    if ((isInitAddr == 0) || (isInitMask == 0))
+                    {
+                        JITDUMP("getIsClassInitedFieldAddress: returned empty isInitAddr or isInitedMask.");
+                        return PhaseStatus::MODIFIED_NOTHING;
```

Comment:
Shouldn't you keep looking for other initializer calls?

Reply:
ah, indeed.